### ISSUE TITLE
Final calc for linear types needs to be divided into abox and ifs

### ISSUE THREAD

**Issue by mclarekin** [buddhi1980/mandelbulber2#424]
the following needs updating. Analytical LINEAR DE needs to be divided into  abox and IFS

// calculate estimated distance

#ifdef IS_HYBRID
#ifdef ANALYTIC_LOG_DE
	dist = 0.5f * aux.r * native_log(aux.r) / (aux.r_dz);
**#elif ANALYTIC_LINEAR_DE
	dist = (aux.r - 2.0f) / fabs(aux.DE);**
#elif ANALYTIC_PSEUDO_KLEINIAN_DE
	float rxy = length(z.xy);
	dist = max(rxy - aux.pseudoKleinianDE, fabs(rxy * z.z) / aux.r) / (aux.DE);
#elif ANALYTIC_JOS_KLEINIAN_DE
	if (consts->fractal[0].transformCommon.functionEnabled)
		z.y = min(z.y, consts->fractal[0].transformCommon.foldingValue - z.y);
	dist = min(z.y, consts->fractal[0].analyticDE.tweak005)
				 / max(aux.pseudoKleinianDE, consts->fractal[0].analyticDE.offset1);

**Comment by buddhi1980** [comment]
The same code is in C++ version:

```
		if (fractals.IsHybrid())
		{
			if (extendedAux.r_dz > 0)
			{
				if (fractals.GetDEFunctionType(0) == fractal::linearDEFunction)
				{
					out->distance = (r - in.common.linearDEOffset) / fabs(extendedAux.DE);
				}
				else if (fractals.GetDEFunctionType(0) == fractal::logarithmicDEFunction)
				{
					out->distance = 0.5 * r * log(r) / extendedAux.r_dz;
				}
				else if (fractals.GetDEFunctionType(0) == fractal::pseudoKleinianDEFunction)
				{
					double rxy = sqrt(z.x * z.x + z.y * z.y);
					out->distance =
						max(rxy - extendedAux.pseudoKleinianDE, fabs(rxy * z.z) / r) / (extendedAux.DE);
				}
				else if (fractals.GetDEFunctionType(0) == fractal::josKleinianDEFunction)
				{
					if (fractals.GetFractal(0)->transformCommon.functionEnabled)
						z.y = min(z.y, fractals.GetFractal(0)->transformCommon.foldingValue - z.y);

					out->distance =
						min(z.y, fractals.GetFractal(0)->analyticDE.tweak005)
						/ max(extendedAux.pseudoKleinianDE, fractals.GetFractal(0)->analyticDE.offset1);
				}
			}
			else
			{
				out->distance = r;
			}
		}
```

Do you want to update both? It was made in this way because this code works quite well with IFS and aBox. It's not a good idea to have one more function in hybrid mode, because it will lead to use deltaDE when two fractals in different slots will have different DE functions. IFS and aBox in hybrids will be rendered using DeltaDE. Now there is used one common function for both, so they work in analytic mode.

**Comment by zebastian** [comment]
let me know if you need any help with manually transpiling cpp to opencl code; also let me know if some parts of the code can be automatically parsed. I think it would be good to automatically generate as much opencl code as possible, this way the code maintenance in the future will be easiest...

**Comment by buddhi1980** [comment]
This part of the code will be difficult to translate automatically, because I use a lot of #define to optimize the code. ComputeFractal() needs mych more care than another parts of the code, because is crucial for rendering speed. Actually we don't any more code which could be automatically parsed from C++ to OpenCL.

**Comment by mclarekin** [comment]
remember in  fractal list .cpp  we have    analyticFunctionIFS
and analyticFunctionLinear

I think in most cases you will see little difference but with some settings
the difference is important???

**Comment by mclarekin** [comment]
I make it clearer, this is normal mode not not hybrid mode

**Comment by mclarekin** [comment]
OK I had them mixed about, now this is the true issue.

The 2.0f should be in.common.linearDEOffset

c++
if (fractals.GetDEFunctionType(0) == fractal::linearDEFunction)
    {
     out->distance = (r - in.common.linearDEOffset) / fabs(extendedAux.DE);
    }

openCl
In hybrid mode
#elif ANALYTIC_LINEAR_DE
 dist = (aux.r - 2.0f) / fabs(aux.DE);

**Comment by mclarekin** [comment]
closed with commit ebb169befa3429511a30d01b779e8791140c4dd1
